File: typescript/packages/bench/src/metrics/npsMetric.ts
/** Stand-in satisfaction score derived from the customer count. */

import { MetricUnit } from "../contracts/metricContract";
import type { Dataset } from "../dataset";
import { AbstractMetric } from "../support/abstractMetric";

/** Stand-in satisfaction score derived from the customer count. */
export class NpsMetric extends AbstractMetric {
    /** Registry key, unique across the lane. */
    static readonly KEY = "nps";

    constructor() {
        super(NpsMetric.KEY, MetricUnit.Count);
    }

    override compute(data: Dataset): number {
        return data.customers.length * 8.5;
    }
}

File: rust/crates/atelier-bench/src/metrics/nps_metric.rs
//! Stand-in satisfaction score derived from the customer count.

use crate::contracts::metric_contract::{MetricContract, MetricUnit};
use crate::dataset::Dataset;
use crate::support::abstract_metric::MetricBase;

/// Stand-in satisfaction score derived from the customer count.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct NpsMetric {
    base: MetricBase,
}

impl NpsMetric {
    /// Registry key, unique across the lane.
    pub const KEY: &'static str = "nps";

    pub const fn new() -> Self {
        Self {
            base: MetricBase::new(Self::KEY, MetricUnit::Count),
        }
    }

    pub const fn base(&self) -> &MetricBase {
        &self.base
    }
}

impl Default for NpsMetric {
    fn default() -> Self {
        Self::new()
    }
}

impl MetricContract for NpsMetric {
    fn key(&self) -> &'static str {
        self.base.key
    }

    fn unit(&self) -> MetricUnit {
        self.base.unit
    }

    fn compute(&self, data: &Dataset) -> f64 {
        data.customers.len() as f64 * 8.5
    }
}
